sw/device/tests/rv_core_ibex_rnd.c

```diff
@@ -0,0 +1,102 @@
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+dif_result_t dif_rv_core_ibex_read_rnd_data_followed_by_status(
+    const dif_rv_core_ibex_t *rv_core_ibex, uint32_t *data,
+    dif_rv_core_ibex_rnd_status_t *followed_status) {
+  if (rv_core_ibex == NULL || data == NULL || followed_status == NULL) {
+    return kDifBadArg;
+  }
+  uint32_t reg = mmio_region_read32(rv_core_ibex->base_addr,
+                                    RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
+  if (!bitfield_bit32_read(reg, RV_CORE_IBEX_RND_STATUS_RND_DATA_VALID_BIT)) {
+    return kDifError;
+  }
+
+  *data = mmio_region_read32(rv_core_ibex->base_addr,
+                             RV_CORE_IBEX_RND_DATA_REG_OFFSET);
+  *followed_status = mmio_region_read32(rv_core_ibex->base_addr,
+                                        RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
+  return kDifOk;
+}
+
+dif_result_t dif_rv_core_ibex_multiple_unchecked_read_rnd_data(
+    const dif_rv_core_ibex_t *rv_core_ibex, const uint32_t repetitions,
+    uint32_t *data) {
+  if (rv_core_ibex == NULL || data == NULL) {
+    return kDifBadArg;
+  }
+
+  for (uint32_t i = 0; i < repetitions; i++) {
+    *data = mmio_region_read32(rv_core_ibex->base_addr,
+                               RV_CORE_IBEX_RND_DATA_REG_OFFSET);
+  }
+  return kDifOk;
+}
+
+#define RANDOM_DATA_READS 16
+
+bool test_main(void) {
+  // Verify the functionality of the random number generation CSRs.
+
+  // Add variables to the stack.
+  dif_rv_core_ibex_t rv_core_ibex;
+  uint32_t rnd_data, previous_rnd_data;
+  dif_rv_core_ibex_rnd_status_t rnd_status = 0;
+  dif_edn_t edn0;
+  dif_edn_seed_material_t seed_material;
+  dif_result_t return_value;
+  const uint32_t unchecked_read_repititions = 64;
+
+  // Enable entropy complex so `RND_DATA` can get entropy.
+  seed_material.data[0] = 0x5a5a5a5a;
+  seed_material.len = 1;
+  CHECK_DIF_OK(
+      dif_edn_init(mmio_region_from_addr(TOP_EARLGREY_EDN0_BASE_ADDR), &edn0));
+  CHECK_DIF_OK(dif_edn_instantiate(&edn0, kDifEdnEntropySrcToggleEnable,
+                                   &seed_material));
+
+  // Perform multiple reads from `RND_DATA` polling `RND_STATUS` in between to
+  // only read valid data. Check different random bits are provided each time.
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &rv_core_ibex));
+
+  for (int i = 0; i < RANDOM_DATA_READS; i++) {
+    while (rnd_status == 0) {
+      CHECK_DIF_OK(dif_rv_core_ibex_get_rnd_status(&rv_core_ibex, &rnd_status));
+    }
+    CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data(&rv_core_ibex, &rnd_data));
+    if (i != 0) {
+      CHECK(rnd_data != previous_rnd_data);
+    }
+    previous_rnd_data = rnd_data;
+  }
+
+  // Ensure `RND_STATUS` indicate invalid data immediately after `RND_DATA`
+  // read.
+  CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data_followed_by_status(
```

Comment:
I think we should change the `dif_rv_core_ibex_read_rnd_data` to not check the status, so it will be more generic and you will not need to create these new functions.